### PR TITLE
feat: exposed rotating-file log for json logging

### PR DESCRIPTION
**Pull request by mlucool**
This both allows for logger type 'rotating-file' and
passing of other options from the config "option" example:
```
{type: rotating-file, format: json, path: /path/to/log.jsonl, level: http, options: {period: 1d}}
```

Resolves: https://github.com/verdaccio/verdaccio/issues/388
Refers: https://github.com/verdaccio/verdaccio/pull/427

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -44,6 +44,7 @@
     "mime": "2.3.1",
     "minimatch": "3.0.4",
     "mkdirp": "0.5.1",
+    "mv": "^2.1.1",
```
Comment (ayusharma) [inline package.json]: Should it not already installed by Logger  ? https://github.com/trentm/node-bunyan/blob/fe31b83e42d9c7f784e83fdcc528a7c76e0dacae/package.json#L27

File: src/lib/logger.js [hunk 2]
```diff
@@ -22,6 +23,16 @@ function getlvl(x) {
   }
 }
 
+/**
+ * A RotatingFileStream that modifes the message first
+ */
+class VerdaccioRotatingFileStream extends Logger.RotatingFileStream { // We depend on mv so that this is there
```
Comment (ayusharma) [inline src/lib/logger.js]: @juanpicado I am sure about extending the `Logger.RotatingFileStream`

File: src/lib/logger.js [hunk 3]
```diff
@@ -46,42 +57,79 @@ function setup(logs) {
...
 
+      if (target.type === 'file') {
+        // destination stream
+        dest = require('fs').createWriteStream(target.path, {flags: 'a', encoding: 'utf8'});
+        dest.on('error', function(err) {
+          stream.emit('error', err);
+        });
+      } else if (target.type === 'stdout' || target.type === 'stderr') {
+        dest = target.type === 'stdout' ? process.stdout : process.stderr;
+        destIsTTY = dest.isTTY;
```
Comment (ayusharma) [inline src/lib/logger.js]: `destIsTTY` it is not used further. ??

File: src/lib/logger.js [hunk 4]
```diff
@@ -46,42 +57,79 @@ function setup(logs) {
...
+          }
+        );
     } else {
-      stream.write = (obj) => {
-        dest.write(jsonPrint(obj));
+      const stream = new Stream();
+      stream.writable = true;
+
+      let dest;
+      let destIsTTY = false;
```
Comment (ayusharma) [inline src/lib/logger.js]: it can be a constant.

File: src/lib/logger.js [hunk 5]
```diff
@@ -46,42 +57,79 @@ function setup(logs) {
...
+      const prettyPrint = (obj) => print(obj.level, obj.msg, obj, destIsTTY) + '\n';
+      const prettyTimestampedPrint = (obj) => obj.time.toISOString() + print(obj.level, obj.msg, obj, destIsTTY) + '\n';
+      const jsonPrint = (obj) => {
+        const msg = fillInMsgTemplate(obj.msg, obj, destIsTTY);
+        return JSON.stringify({...obj, msg}, Logger.safeCycles()) + '\n';
       };
-    }
 
+      if (target.type === 'file') {
```
Comment (ayusharma) [inline src/lib/logger.js]: define `dest` here. I'll suggest using full name instead of short forms. `destination`

File: package.json (hunk 1 above)
Comment (mlucool) [reply package.json]: It's optional and this forces it to be there. Their code does if(mv): https://github.com/trentm/node-bunyan/blob/master/lib/bunyan.js#L96

File: src/lib/logger.js (hunk 2 above)
Comment (mlucool) [reply src/lib/logger.js]: What do you suggest instead? Reimplementing it is a lot of work and this is public interface so a breaking change would be declarative

File: src/lib/logger.js [hunk 6]
```diff
@@ -22,6 +23,16 @@ function getlvl(x) {
   }
 }
 
+/**
+ * A RotatingFileStream that modifes the message first
+ */
+class VerdaccioRotatingFileStream extends Logger.RotatingFileStream { // We depend on mv so that this is there
+  write(obj) {
+    const msg = fillInMsgTemplate(obj.msg, obj, false);
```
[the remark this answers is not in the document]
Comment (mlucool) [reply src/lib/logger.js]: https://github.com/verdaccio/verdaccio/blob/ef02b2ad5119b53ce9f5c127bb67a672ca73f1f8/src/lib/logger.js#L187

File: src/lib/logger.js (hunk 3 above)
Comment (mlucool) [reply src/lib/logger.js]: It is used above, which is why those functions are in this scope. It's basically binding them all to the variables in this block scope

File: src/lib/logger.js [hunk 7]
```diff
@@ -46,42 +57,79 @@ function setup(logs) {
       return JSON.stringify({...obj, msg}, Logger.safeCycles()) + '\n';
     };
 
-    if (target.type === 'file') {
```
Comment (juanpicado) [inline src/lib/logger.js]: From line 53 to 58, the following code is duplicated below and not being unused.
```
    const prettyPrint = (obj) => print(obj.level, obj.msg, obj, destIsTTY) + '\n';
    const prettyTimestampedPrint = (obj) => obj.time.toISOString() + print(obj.level, obj.msg, obj, destIsTTY) + '\n';
    const jsonPrint = (obj) => {
      const msg = fillInMsgTemplate(obj.msg, obj, destIsTTY);
      return JSON.stringify({...obj, msg}, Logger.safeCycles()) + '\n';
    };
```

File: package.json (hunk 1 above)
Comment (juanpicado) [reply package.json]: Remove the `^` please

File: src/lib/logger.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (ayusharma) [reply src/lib/logger.js]: Sorry, I forgot to update my master branch.

**Comment by juanpicado** [comment]
````
Found 0 errors

/Users/user/projects/verdaccio/src/lib/logger.js
  51:9   error  'dest' is defined but never used                             no-unused-vars
  53:11  error  'prettyPrint' is assigned a value but never used             no-unused-vars
  54:11  error  'prettyTimestampedPrint' is assigned a value but never used  no-unused-vars
  55:11  error  'jsonPrint' is assigned a value but never used               no-unused-vars

✖ 7 problems (4 errors, 3 warnings)
```

**Comment by juanpicado** [comment]
⚠️ Cluster mode
> WARNING on cluster usage: Using Bunyan's rotating-file stream with node.js's "cluster" module can result in unexpected file rotation. You must not have multiple processes in the cluster logging to the same file path. In other words, you must have a separate log file path for the master and each worker in the cluster. Alternatively, consider using a system file rotation facility such as logrotate on Linux or logadm on SmartOS/Illumos. See this comment on issue #117 for details.

https://github.com/trentm/node-bunyan#stream-type-rotating-file

@mlucool that deserves to be in the docs as well. As a reference at the bottom below the table

File: src/lib/logger.js (hunk 3 above)
Comment (ayusharma) [reply src/lib/logger.js]: You can pass `destIsTTY` in pretty `prettyTimestampedPrint`, `jsonPrint` explicitily. Currently it is not readable.

**Comment by juanpicado** [comment]
It seems does not work, it fills files partially and then are emtpy. 
```
# log settings
logs:
#  - {type: stdout, format: pretty, level: http}
#  - {type: file, path: /Users/jpicado/verdaccio.log, level: trace}
  - {type: rotating-file, format: json, path: /Users/jpicado/verdaccio-rotate.json, level: http, options: {period: 1000ms}}
```

ezgif com-video-to-gif 4

File: src/lib/logger.js (hunk 2 above)
Comment (ayusharma) [reply src/lib/logger.js]: make sense.

**Comment by mlucool** [comment]
@juanpicado @ashishsurana Thanks for the comments, I'll look at this next week.

> It seems does not work, it fills files partially and then are emtpy.

That is expected. It is a time based rotate only and will rotate even if there is no new data. In your example, you have it rotating every 1s. Try navigating the website or installing something every second and you'll see that logs continue to have data.

File: src/lib/logger.js (hunk 7 above)
Comment (mlucool) [reply src/lib/logger.js]: Meh, bad rebase on my part. Will fix

File: src/lib/logger.js (hunk 3 above)
Comment (mlucool) [reply src/lib/logger.js]: Ok, I'll refactor this class. Was trying to avoid that, but it is now unreadable, its time to refactor :)

Comment (juanpicado) [reply src/lib/logger.js]: Thanks @mlucool feel free to refactor the whole `logger.js` if you wish

**Comment by juanpicado** [comment]
If there are no changes on `logger` configuration, it will be shipped with next minor `v3.8.0`release.

**Comment by mlucool** [comment]
@juanpicado @ayusharma please re-review.

**Comment by juanpicado** [comment]
@mlucool I was on it but .... sadly my ISP decided I have to change my router 😢  ... I'll be offline rest of the week until fix that.

File: yarn.lock [hunk 8]
```diff
@@ -6647,7 +6647,7 @@ mute-stream@0.0.7:
   version "0.0.7"
   resolved "https://registry.npmjs.org/mute-stream/-/mute-stream-0.0.7.tgz#3075ce93bc21b8fab43e1bc4da7e8115ed1e7bab"
 
-mv@~2:
+mv@^2.1.1, mv@~2:
```
Comment (juanpicado) [inline yarn.lock]: `mv@^2.1.1, mv@~2:` <-- the `^` remains in lock file

**Comment by juanpicado** [comment]
Thanks @mlucool ! great job

**Comment by juanpicado** [comment]
@mlucool fyi
https://github.com/verdaccio/verdaccio/releases/tag/v3.8.0

**Comment by mlucool** [comment]
Thanks!

**Pull request merged**